sensor/common/registry/registry_store.go

```diff
@@ -392,3 +392,233 @@ func (rs *Store) GetMatchingCentralRegistryIntegrations(imgName *storage.ImageNa
+	}
+
+	hostToRegistry[registryAddr] = reg
+}
+
+// DeleteSecret returns true when a secret is deleted from the store, false otherwise.
+func (rs *Store) DeleteSecret(namespace, secretName string) bool {
+	if !features.SensorPullSecretsByName.Enabled() {
+		// When storing secrets by host they cannot be deleted.
+		return false
+	}
+
+	rs.storeMutux.Lock()
+	defer rs.storeMutux.Unlock()
+
+	secretNameToHost := rs.storeByName[namespace]
+	if secretNameToHost == nil {
+		return false
+	}
+
+	if _, ok := secretNameToHost[secretName]; ok {
+		delete(secretNameToHost, secretName)
+
+		if len(secretNameToHost) == 0 {
+			// If there are no more secrets for this namespace, delete the namespace entry as well.
+			delete(rs.storeByName, namespace)
+		}
+
+		log.Debugf("Deleted secret %q from namespace %q", secretName, namespace)
+		return true
+	}
+
+	return false
+}
+
+// GetPullSecretRegistries returns the matching registries associated with the provided pull secrets found in namespace.
+// If no pull secrets are provided, all matching registries from the namespace are returned.
+func (rs *Store) GetPullSecretRegistries(image *storage.ImageName, namespace string, imagePullSecrets []string) ([]types.ImageRegistry, error) {
+	if !features.SensorPullSecretsByName.Enabled() {
+		reg, err := rs.getRegistryForImageInNamespace(image, namespace)
+		if err != nil {
+			return nil, err
+		}
+
+		return []types.ImageRegistry{reg}, nil
+	}
+
+	rs.storeMutux.RLock()
+	defer rs.storeMutux.RUnlock()
+
+	secretNameToHost, ok := rs.storeByName[namespace]
+	if !ok {
+		return nil, nil
+	}
+
+	if len(imagePullSecrets) > 0 {
+		// Return matching registries referenced by the image pull secrets.
+		return rs.getPullSecretRegistriesNoLock(secretNameToHost, image, imagePullSecrets), nil
+
+	}
+
+	// If no pull secrets were provided, we assume that all matching registries
+	// from the namespace are desired (scan requests that originate from Central
+	// will not have pull secrets, such as those executed via roxctl).
+	return rs.getAllPullSecretRegistriesNoLock(secretNameToHost, image), nil
+}
+
+// getPullSecretRegistriesNoLock returns registries found within image pull secrets
+// from a namespace that match image.
+func (rs *Store) getPullSecretRegistriesNoLock(secretNameToHost secretNameToHostname, image *storage.ImageName, imagePullSecrets []string) []types.ImageRegistry {
+	var regs []types.ImageRegistry
+	registryHostname := image.GetRegistry()
+
+	// Extract registries from the matching pull secrets.
+	for _, secretName := range imagePullSecrets {
+		for host, reg := range secretNameToHost[secretName] {
+			if host == registryHostname {
+				regs = append(regs, reg)
+			}
+		}
+	}
+
+	return regs
+}
+
+// getAllPullSecretRegistriesNoLock returns all registries within a namespace that match image.
+func (rs *Store) getAllPullSecretRegistriesNoLock(secretNameToHost secretNameToHostname, image *storage.ImageName) []types.ImageRegistry {
```

Comment:
[nit] This function could simply call `getPullSecretRegistriesNoLock` after gathering all the secret names (line 613) right?

Reply:
Yes! Good call / missed that.